Feature/git smart http

oliverguenther opened the pull request:
(This is https://github.com/opf/openproject/pull/1629, refactored and updated for dev

This PR extends `OpenProjectAuthentication.pm` to allow Git Smart HTTP of repositories from OpenProject with access through Apache, while authorization is largely identical to the subversion repositories.
### Original PR:

I'd like to see git supported in our authentication scripts (just like we support svn).

This PR aims to do that.

see: https://www.openproject.org/work_packages/14383

tessi committed:
Add git support to repository authentication script

Refactors the original PR from #1629
* dispersed authorization methods in SysController
* refactored and updated SysController specs

oliverguenther committed:
Extract SCM authorization routines into policies per vendor

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-git-smart-http.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-git-smart-http.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

tessi commented:
Thanks @oliverguenther for keeping up with integrating this :) :clap:

oliverguenther committed:
Resolve LoadError for scm policy

ulferts commented:
Hey @tessi, care to do the review? Somebody has to and you seem to be the right man for the job.

ulferts commented on spec/controllers/sys_controller_spec.rb:
```diff
@@ -28,208 +28,549 @@
 
 require 'spec_helper'
 
-module OpenProjectRepositoryAuthenticationSpecs
```
You should either turn this into a request/request spec or considerably reduce the stack you are testing. You are currently not unit testing as you are testing down to the model.

I'd:
- write request specs for the happy case
- write unit test for the controller where you mock like hell
- write unit tests for the policies

That should reduce the amount of testing you are doing.

oliverguenther replied:
I agree. But I'd prefer to merge this and refactor the specs in a separate PR.

ulferts commented:
Code itself is looking good but I admit to not having tested it as I have no apache set up and ready for testing.

oliverguenther commented:
I have tested it myself on Apache inside a VM, however I would appreciate someone validating my findings.

I'll also take another look later today concerning the functionality after Rails 4 merge.

oliverguenther committed:
Use explicit route for repo_auth

oliverguenther committed:
Use --shared to create bare Git repositories

Using Git --shared sets default permissions to group writable,
otherwise when the user running OpenProject differs from the webserver
serving Git (Apache in the supported setup), the permissions for changed
files through Git Smart HTTP cause the repository to be no longer
deletable by the OpenProject runuser.

tessi commented:
Code-wise this looks good. I don't have an apache OP setup handy. Back when I've created the first commit, the thing worked, though.

All changes by @oliverguenther since then seem to make things better. So this has my :+1:

oliverguenther commented:
I've deployed this on my Apache VM and found two minor issues caused by the changes from Rails 4, but otherwise this worked fine. I'm still a little scared regarding that Git read/only check (especially regarding implementation changes in that regard), but I don't think we can make it more explicit on that side.

tessi commented:
Thanks again @oliverguenther for keeping track of this. I'm happy to see this merged
